### PR TITLE
Operator overloading for filter syntax

### PR DESCRIPTION
**Pull request by benbc**

### CODE REVIEW

**Comment by benbc** [comment]
@inglesp A few things about the surface here that deserve discussion.

1. `flake8` doesn't like us doing equality comparisons with `True`. I've experimented here with allowing `BoolColumn`s to be used directly as predicates to cope with that case. Let me know what you think of that.
2. `flake8` _also_ doesn't like doing equality comparisons against `None`. I've silenced it in that instance. I think there's a case for allowing _any_ column to be passed "naked" in order express a null check (like `events.filter(events.code)` instead of `events.filter(events.code != None)`). Not sure it's clear, though. I'd like to understand the use case for these checks.
3. I think maybe we should invert the codelist thing, so that we do `events.filter(code.is_in(codelist))` rather than `events.filter(codelist.contains(code))`. Seems a bit more consistent with the operator overloading elsewhere and it's a possibility now that we have these column types where before it was not possible.

**Comment by benbc** [comment]
@inglesp I've added a commit which inverts the codelist membership test. Easy to pull it out if you don't like it.

**Comment by benbc** [comment]
@inglesp And another commit with an alternative approach for checking nulls.

File: cohortextractor2/dsl.py [hunk 1]
```diff
@@ -221,10 +213,51 @@ def __hash__(self) -> int:
...
+    def is_true(self) -> Predicate:
+        return Predicate(self, "equals", True)
+
+
+class DateColumn(Column):
+    def __gt__(self, other: str) -> Predicate:
+        return Predicate(self, "greater_than", other)
+
+    def __lt__(self, other: str) -> Predicate:
+        return Predicate(self, "less_than", other)
```
Comment (rebkwok) [inline cohortextractor2/dsl.py:250]: Should this have other operators too?  It's reasonable to expect to be able do <=, >=, ==, != on dates.  Possibly it'd make sense for there to be a base Column for shared operator methods - I'd expect `DateColumn` and `IntColumn` at least to share some of these methods.

File: cohortextractor2/query_language.py [hunk 2]
```diff
@@ -1,3 +1,5 @@
+from __future__ import annotations
```
Comment (rebkwok) [inline cohortextractor2/query_language.py:1]: There was some discussion about whether this was a bad idea or not;  I'm not sure we came to any conclusions, but maybe it should be discussed some more

File: tests/test_dsl.py [hunk 3]
```diff
@@ -67,7 +68,7 @@ class OldCohort(OldCohortWithPopulation):
     cohort = cohort_with_population
     events = tables.clinical_events
     cohort.code = (
-        events.filter(codelist(["Code1"], "ctv3").contains(events.code))
+        events.filter(events.code.is_in(codelist(["Code1"], "ctv3")))
```
Comment (rebkwok) [inline tests/test_dsl.py:71]: :+1: I prefer this

File: cohortextractor2/dsl.py (hunk 1 above)
Comment (rebkwok) [reply cohortextractor2/dsl.py:250]: Also a date-specific `betweeen` method?

**Review by rebkwok — APPROVED** [review]
LGTM.  A few comments/questions but nothing major

File: cohortextractor2/dsl.py [hunk 4]
```diff
@@ -91,11 +90,7 @@ class EventFrame:
...
 
-    def filter(  # noqa: A003
-        self,
-        column_or_expr: Column | CodelistFilterExpr,
-        **kwargs: str | Codelist | None | bool,
-    ) -> EventFrame:
+    def filter(self, predicate: Predicate | BoolColumn) -> EventFrame:  # noqa: A003
         """Return a new EventFrame with given filter.
 
         Note that while we are building the DSL, this method takes either an expression
```
[the remark this answers is not in the document]
Comment (benbc) [reply cohortextractor2/dsl.py:96]: Oh yes, I meant to come back to that. Thanks.

File: cohortextractor2/dsl.py (hunk 1 above)
Comment (benbc) [reply cohortextractor2/dsl.py:250]: Oh yes, absolutely. I just added enough to satisfy the existing tests. You're right that I should chuck in a bunch more while I'm at it.

Comment (benbc) [reply cohortextractor2/dsl.py:250]: > Possibly it'd make sense for there to be a base Column for shared operator methods

Yes, I wondered about that. Or about having mixins for groups of operators. Or something. I'll try filling out the set of operators and then see if it makes sense to refactor.

File: cohortextractor2/dsl.py [hunk 5]
```diff
@@ -222,6 +214,18 @@ def __hash__(self) -> int:
         return hash(repr(self.value))
 
 
+class Predicate:
+    def __init__(
+        self, column: Column, operator: str, other: str | bool | Codelist | None
```
Comment (inglesp) [inline cohortextractor2/dsl.py:218]: Must the LHS of the predicate be a `Column`?

This means we can do `.filter(col > 123)` but can't do `.filter(123 < col)`.

File: tests/test_dsl.py (hunk 3 above)
Comment (inglesp) [reply tests/test_dsl.py:71]: Me too.

Also, we can return to using `make_codelist` in these tests.

File: cohortextractor2/dsl.py (hunk 5 above)
Comment (inglesp) [reply cohortextractor2/dsl.py:218]: Also: we're going to want to use `PatientSeries` in predicates.

For instance, in the example I documented we've got `covid_imms.filter(imms.date > (cohort.first_date + 28))`.  `imms.date` is a `Column`, but `cohort.first_date` and `cohort.first_date + 28` are `PatientSeries`.

**Review by inglesp — APPROVED** [review]
All looks good and easy to follow.

To respond to your questions:

> I've experimented here with allowing `BoolColumns` to be used directly as predicates to cope with that case...

Makes good sense.  We'd want to catch users doing `col == True` and tell them what they should do instead.  Also: will we need a way to negate a column to provide an alternative to `.filter(col == False)`?

> I think there's a case for allowing any column to be passed "naked" in order express a null check (like `events.filter(events.code)` instead of `events.filter(events.code != None)`)...

I like the explicit `isnull` that you've added in 6b2cb5a.

> I think maybe we should invert the codelist thing...

Agree.

---

Happy for this to be merged soon, but I'd like to revisit using `PatientSeries` in predicates.

**Comment by benbc** [comment]
@inglesp 

> > I've experimented here with allowing BoolColumns to be used directly as predicates to cope with that case...

> Makes good sense. We'd want to catch users doing col == True and tell them what they should do instead.

Or, it occurs to me as I implement it, we could just allow them to do the `== True` check -- but provide the alternative `.is_true()`, so that they have a way of doing it that doesn't cause any warnings. WDYT?

**Comment by inglesp** [comment]
> @inglesp
> 
> > > I've experimented here with allowing BoolColumns to be used directly as predicates to cope with that case...
> 
> > Makes good sense. We'd want to catch users doing col == True and tell them what they should do instead.
> 
> Or, it occurs to me as I implement it, we could just allow them to do the `== True` check -- but provide the alternative `.is_true()`, so that they have a way of doing it that doesn't cause any warnings. WDYT?

I don't have a strong opinion.  I'm happy with supporting multiple ways of doing something, as long as they're properly equivalent.

File: cohortextractor2/dsl.py (hunk 5 above)
Comment (benbc) [reply cohortextractor2/dsl.py:218]: It seems that writing the comparison the other way round just works (TM).

Comment (benbc) [reply cohortextractor2/dsl.py:218]: I've added a separate story for using patient series in predicates.

**Pull request merged**